### PR TITLE
feature: add publish-all cli flag

### LINKED ISSUE #2517: [feature-request] support `publishAll` flag in container's command line
### Ⅰ. Issue Description

In the https://github.com/alibaba/pouch/pull/2500#discussion_r236930266, we discussed about the flag `publishAll` or `publicAll`.

In https://github.com/alibaba/pouch/blob/master/cli/container.go#L68, we have field `publicAll      bool`.

But I am afraid that we have not used this field in anywhere.

However, in Moby/Docker, we have https://github.com/docker/cli/blob/2634562119dd5b4e64d529d1aa5e16f760382d5f/cli/command/container/opts.go#L69:
```
publishAll         bool
```
And the flag : 
```
flags.BoolVarP(&copts.publishAll, "publish-all", "P", false, "Publish all exposed ports to random ports")
```

### Ⅱ. Describe what happened
no support `publish-all` flag

### Ⅲ. Describe what you expected to happen
support `publish-all` flag

### PR DESCRIPTION
**Pull request by ZYecho**
Signed-off-by: zhangyue <zy675793960@yeah.net>

### Ⅰ. Describe what this PR did
add cli publish all flag to support publish all expose port.

### Ⅱ. Does this pull request fix one issue?

fix #2517 

### Ⅲ. Why don't you add test cases (unit test/integration test)? (你真的觉得不需要加测试吗？)
add a integration test.

### Ⅳ. Describe how to verify it
`pouch run -d -P redis`
to get a publish all container.
`pouch inspect cefab5`
Got the random port
```
 "NetworkSettings": {
            "Networks": {
                "bridge": {
                    "Aliases": null,
                    "EndpointID": "3cf3f181d5066cc607a64faaf1149fe2745fc607781cd071c499aa8fbf696931",
                    "Gateway": "192.168.5.1",
                    "IPAddress": "192.168.5.2",
                    "IPPrefixLen": 24,
                    "Links": null,
                    "MacAddress": "02:42:c0:a8:05:02",
                    "NetworkID": "12541cab332aaabf2695006c547441d38b46fb94f857db65a491165acedd1518"
                }
            },
            "Ports": {
                "6379/tcp": [
                    {
                        "HostIp": "0.0.0.0",
                        "HostPort": "32768"
                    }
                ]
            },
            "SandboxID": "5f2e05d9b731ac979191c6c8340dbccaa68a784ccaf1e735cc6ea4e07aff2a54",
            "SandboxKey": "/var/run/pouch/netns/5f2e05d9b731",
            "SecondaryIPAddresses": null,
            "SecondaryIPv6Addresses": null
        },

```
use `lsof` to verify
`lsof -iTCP | grep pouch`
Got
`pouchd    15090            root   16u  IPv6 265330      0t0  TCP *:32768 (LISTEN)`

### Ⅴ. Special notes for reviews
None.

### CODE REVIEW

File: daemon/mgr/network.go [hunk 1]
```diff
@@ -744,3 +744,62 @@ func joinOptions(endpoint *types.Endpoint) ([]libnetwork.EndpointOption, error)
...
+		// It is not an error for epInfo to be nil
+		return pm, nil
+	}
+
+	if expData, ok := driverInfo[netlabel.ExposedPorts]; ok {
+		if exposedPorts, ok := expData.([]networktypes.TransportPort); ok {
+			for _, tp := range exposedPorts {
+				natPort, err := nat.NewPort(tp.Proto.String(), strconv.Itoa(int(tp.Port)))
+				if err != nil {
+					return pm, fmt.Errorf("error parsing Port value(%v):%v", tp.Port, err)
```
Comment (allencloud) [inline daemon/mgr/network.go]: `failed to parse Port value(%v): %v`

according to https://github.com/alibaba/pouch/blob/master/docs/contributions/code_styles.md#rule005---generation-of-action-failure

File: daemon/mgr/network.go [hunk 2]
```diff
@@ -744,3 +744,62 @@ func joinOptions(endpoint *types.Endpoint) ([]libnetwork.EndpointOption, error)
...
+	for _, ep := range sb.Endpoints() {
+		pm, _ = getEndpointPortMapInfo(ep)
+		if len(pm) > 0 {
+			break
+		}
+	}
+	return pm
+}
+
+func getEndpointPortMapInfo(ep libnetwork.Endpoint) (apitypes.PortMap, error) {
```
Comment (allencloud) [inline daemon/mgr/network.go]: Could you add more annotation for this function and function `getSandboxPortMapInfo`? Actually I could get the meaning of both functions. In addition, inside of function, some separated code blocks still need some comments for guidance.

File: daemon/mgr/network.go [hunk 3]
```diff
@@ -744,3 +744,63 @@ func joinOptions(endpoint *types.Endpoint) ([]libnetwork.EndpointOption, error)
...
+		}
+	}
+
+	mapData, ok := driverInfo[netlabel.PortMap]
+	if !ok {
+		return pm, nil
+	}
+
+	// get the port-mapping from driverInfo, which was open by -p HostPort:port
+	if portMapping, ok := mapData.([]networktypes.PortBinding); ok {
```
Comment (allencloud) [inline daemon/mgr/network.go]: You could use https://github.com/alibaba/pouch/blob/master/docs/contributions/code_styles.md#rule006---return-fast-to-indent-less to indent less.

```
portMapping, ok := mapData.([]networktypes.PortBinding)
if !ok {
    return pm, nil
}
for _, pp := range portMapping {
      ......
}
return pm, nil
```

File: test/cli_run_network_test.go [hunk 4]
```diff
@@ -33,3 +35,18 @@ func (suite *PouchRunNetworkSuite) TestRunWithPing(c *check.C) {
 	defer DelContainerForceMultyTime(c, name)
 	res.Assert(c, icmd.Success)
 }
+
+// TestRunWithPublicAll is to verify run container with publish-all flag
+func (suite *PouchRunNetworkSuite) TestRunWithPublishAll(c *check.C) {
```
Comment (allencloud) [inline test/cli_run_network_test.go]: Could you also add a test case for testing an image which expose multiple ports? @ZYecho

Comment (ZYecho) [reply test/cli_run_network_test.go]: ok, and could you help to check to CI result? it looks strange and I don't know why it failed.

**Comment by allencloud** [comment]
This is the only CI failure in integration test:
```
----------------------------------------------------------------------
FAIL: /home/travis/gopath/src/github.com/alibaba/pouch/test/cli_inspect_test.go:195: PouchInspectSuite.TestContainerInspectPorts
/home/travis/gopath/src/github.com/alibaba/pouch/test/cli_inspect_test.go:208:
    c.Assert(string(data), check.Equals, "{\"80/tcp\":[{\"HostPort\":\"8080\"}]}")
... obtained string = "{\"80/tcp\":[{\"HostIp\":\"0.0.0.0\",\"HostPort\":\"8080\"}]}"
... expected string = "{\"80/tcp\":[{\"HostPort\":\"8080\"}]}"
----------------------------------------------------------------------
```
@ZYecho  I think it is related to the code change.

**Comment by ZYecho** [comment]
> This is the only CI failure in integration test:
> 
> ```
> ----------------------------------------------------------------------
> FAIL: /home/travis/gopath/src/github.com/alibaba/pouch/test/cli_inspect_test.go:195: PouchInspectSuite.TestContainerInspectPorts
> /home/travis/gopath/src/github.com/alibaba/pouch/test/cli_inspect_test.go:208:
>     c.Assert(string(data), check.Equals, "{\"80/tcp\":[{\"HostPort\":\"8080\"}]}")
> ... obtained string = "{\"80/tcp\":[{\"HostIp\":\"0.0.0.0\",\"HostPort\":\"8080\"}]}"
> ... expected string = "{\"80/tcp\":[{\"HostPort\":\"8080\"}]}"
> ----------------------------------------------------------------------
> ```
> @ZYecho I think it is related to the code change.

Thanks, fixed it.

File: test/cli_run_network_test.go (hunk 4 above)
Comment (ZYecho) [reply test/cli_run_network_test.go]: done.

File: daemon/mgr/container.go [hunk 5]
```diff
@@ -650,6 +650,15 @@ func (mgr *ContainerManager) prepareContainerNetwork(ctx context.Context, c *Con
 		}
 	}
 
+	sb, err := mgr.NetworkMgr.Controller().SandboxByID(c.NetworkSettings.SandboxID)
+	if err != nil {
+		//sandbox not found, maybe caused by disconnect network or no endpoint
```
Comment (rudyfly) [inline daemon/mgr/container.go]: add space before sandbox, thx.

File: daemon/mgr/container.go [hunk 6]
```diff
@@ -650,6 +650,15 @@ func (mgr *ContainerManager) prepareContainerNetwork(ctx context.Context, c *Con
...
+		logrus.Warnf("failed to get sandbox by id(%s)", c.NetworkSettings.SandboxID)
```
Comment (rudyfly) [inline daemon/mgr/container.go]: I think we should print error information.

File: daemon/mgr/container.go [hunk 7]
```diff
@@ -1461,7 +1470,7 @@ func (mgr *ContainerManager) Disconnect(ctx context.Context, containerName, netw
 	c.Unlock()
 	endpoint.EndpointConfig = epConfig
 	if err := mgr.NetworkMgr.EndpointRemove(ctx, endpoint); err != nil {
-		// TODO(ziren): it is a trick, we should wrapper sanbox
+		// TODO(ziren): it is a trick, we should wrapper sandbox
```
Comment (rudyfly) [inline daemon/mgr/container.go]: 👍

File: test/cli_run_network_test.go [hunk 8]
```diff
@@ -33,3 +36,41 @@ func (suite *PouchRunNetworkSuite) TestRunWithPing(c *check.C) {
...
+	output := command.PouchRun("inspect", name).Stdout()
+	containers := make([]types.ContainerJSON, 1)
+	err := json.Unmarshal([]byte(output), &containers)
+	if err != nil || len(containers) == 0 {
+		c.Fatal("fail to format container json")
+	}
+	c.Assert(len(containers[0].NetworkSettings.Ports), check.Equals, 1)
+	checkPortMapExists(c, containers[0].NetworkSettings.Ports, "8080/tcp")
+
+	// multiple expose port case
```
Comment (rudyfly) [inline test/cli_run_network_test.go]: excellent testcase design.

File: daemon/mgr/container.go (hunk 6 above)
Comment (ZYecho) [reply daemon/mgr/container.go]: There's a scenario in pouch, the container disconnect the network and remove the endpoint, and if there's no endpoint in sandbox, the sandbox will be deleted. So in this scenario, no sandbox in container is not an error?

File: daemon/mgr/container.go (hunk 5 above)
Comment (allencloud) [reply daemon/mgr/container.go]: Any update on this part? @ZYecho

Comment (ZYecho) [reply daemon/mgr/container.go]: done.

File: daemon/mgr/container.go (hunk 6 above)
Comment (rudyfly) [reply daemon/mgr/container.go]: I mean we should print `err`, like this: 
```
logrus.Warnf("failed to get sandbox by id(%s), err(%v)", c.NetworkSettings.SandboxID, err)
```

Comment (ZYecho) [reply daemon/mgr/container.go]: get it, thanks.

**Comment by rudyfly** [comment]
LGTM

**Comment by allencloud** [comment]
Thanks a lot for your work. @ZYecho 
Merging...

**Pull request merged**